python/mspasspy/db/database.py

```diff
@@ -520,30 +520,33 @@ def _sync_metadata_before_update(self, d):
         else:
             d.put_string('time_standard', 'UTC')
 
-    def _save_history(self, mspass_object, history_object_id=None, collection=None):
+    def _save_history(self, mspass_object, prev_history_object_id=None, collection=None):
         """
         Save the processing history of a mspasspy object.
 
         :param mspass_object: the target object.
         :type mspass_object: either :class:`mspasspy.ccore.seismic.TimeSeries` or :class:`mspasspy.ccore.seismic.Seismogram`
-        :param history_object_id: the previous history object id (if it has).
-        :type history_object_id: :class:`bson.objectid.ObjectId`
+        :param prev_history_object_id: the previous history object id (if it has).
+        :type prev_history_object_id: :class:`bson.objectid.ObjectId`
         :param collection: the collection that you want to store the history object. If not specified, use the defined
         collection in the schema.
-        :return: updated history_object_id.
+        :return: current history_object_id.
         """
         if not collection:
             collection = self.database_schema.default_name('history_object')
         history_col = self[collection]
-        history_binary = pickle.dumps(ProcessingHistory(mspass_object))
-        # todo jobname jobid
-        if history_object_id:
+        proc_history = ProcessingHistory(mspass_object)
+        current_uuid = proc_history.id() # uuid in the current node
+        history_binary = pickle.dumps(proc_history)
+        # todo save jobname jobid when global history module is done
+        if prev_history_object_id:
             # overwrite history
-            filter_ = {'_id': history_object_id}
-            history_col.find_one_and_replace(filter_, {'nodesdata': history_binary})
-            return history_object_id
+            history_col.delete_one({'_id': prev_history_object_id})
+            history_col.insert_one({'_id': current_uuid, 'nodesdata': history_binary})
         else:
-            return history_col.insert_one({'nodesdata': history_binary}).inserted_id
+            # new insertion
+            history_col.insert_one({'_id': current_uuid, 'nodesdata': history_binary})
```

Comment:
Is the duplicate key error thrown by this line? I think that's potentially a bug because it is possible for a user to deep copy an object. Maybe we should add an exception handling here such that when the UUID is duplicated, we generate a new uuid. I guess we need to make sure the change is also applied to the `mspass_object` outside of this database operation. Do you think it is gonna work?

Reply:
UUID is designed to be unique on the application-level right?

Reply:
Yes. It is unique, so the problem should only happen when a deep copy is make and the `update_metadata` is called as a new insertion.

Reply:
Sorry, I'm a little confused here. First, why new insertion is related here? Do you mean that when it's a new insertion, we won't delete the old object, so the duplicate error can happen? Then, I do think it's the user's problem to cause a duplicate key error since our processing history is designed like this. Deepcopy should never be recommended in users' case, I'm using it just for the convenience of testing. From another perspective, is it really a good practice that we change the _id implicitly when we find it may produce a duplicate error?

Reply:
Good point! Let's just wrap the error with MsPASSError as you suggested. This should be fine here because no database operation is performed yet at this point.